[영화 극장 선택] 이든 1, 2단계 미션 제출합니다.

devfeijoa opened the pull request:
# 셀프 체크리스트

안녕하세요 제이든 ! 이든입니다.
귀중한 시간 내어 리뷰해주셔서 감사합니다 !

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [ ] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

처음 배우는 개념이 많아서 해당 개념을 공부하고 적용하는데 시간이 많이 걸렸습니다.
그 과정에서 UI테스트, 도메인 테스트를 작성하지 못 했고, step2 미션을 완성하지 못했습니다. 😢 
추후에 리뷰 요청 드릴때 빠르게 반영하여 보내도록 하겠습니다 ! 

도메인을 구성하는 과정에서 영화와 영화관이 다대다 관계를 구성하기 위해서 Screening 이라는 객체를 새로 만들어 그 객체가 영화와 영화관을 받게 했습니다.

현재 예매 결과를 받아오는 화면에서 좌석정보를 강제로 toString 하여 숫자로 받아오고 있는 부분과 데이터 바인딩이 부분적으로 적용되어있는점 인지하고 있습니다 😢 

또한 그 전 단계에서 구현했던 예매화면의 스피너에서 현재 시간 반영하도록 하여, 현재 시간 이후에 상영시간만 뜨도록 하는 기능이 구현되어있지만 사용하고 있지않아 반영되지 않은 점 인지하고 있습니다. 😢 
해당 부분 인지하고 있는만큼 다음 리뷰때 수정하여 보내겠습니다 !


 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

oungsi2000 committed:
feat: setUp

devfeijoa committed:
docs: README 업데이트

devfeijoa committed:
docs: 기능 명세서 작성

oungsi2000 committed:
refactor(ScreeningPeriod): 상영 시간 정책 변경으로 인한 로직 변경

oungsi2000 committed:
refactor: movie 구조 변경에 따른 전체 구조 임시 변경

oungsi2000 committed:
refactor: 상영 객체로 변경

ScreeningPeriod -> Screening

oungsi2000 committed:
feat: 상영 극장 이름, 각 극장의 상영 시간 개수 반환 로직 구현

devfeijoa committed:
refactor: Screening으로 변경

- 컴파일 가능하게 수정

devfeijoa committed:
feat: data binding 설정 추가

devfeijoa committed:
feat: CinemaSeclectionFragment.xml 생성

oungsi2000 committed:
feat: CinemaSeclectionFragment 구현

oungsi2000 committed:
feat: OnCinemaSelectionListener 구현

oungsi2000 committed:
feat: CinemaSelectionAdapter 구현

oungsi2000 committed:
refactor: 패키지 구조 변경

devfeijoa committed:
fix: fragment가 나타나지 않는 문제 수정

devfeijoa committed:
refactor: 극장 선택 xml 수정

devfeijoa committed:
refactor: activity_reservation_result.xml 수정

devfeijoa committed:
refactor: fragment_cinema_seclection.xml 수정

devfeijoa committed:
refactor: step1 구현

devfeijoa committed:
refactor: 객체 역할 변경으로 인한 수정

Movie -> Screening

oungsi2000 committed:
docs:README 업데이트

oungsi2000 committed:
build: data binding 의존성 추가

oungsi2000 committed:
feat: BindingAdapter 생성

oungsi2000 committed:
refactor: movie 데이터 바인딩 적용

oungsi2000 committed:
refactor: cinema 데이터 바인딩 적용

oungsi2000 committed:
refactor: 데이터 바인딩 셋업

oungsi2000 committed:
refactor: Ktlint 적용

oungsi2000 committed:
fix: JVMInline 제거

devfeijoa committed:
refactor: ReservationInfo 파라미터 변경

devfeijoa committed:
refactor: 좌석 정보, 극장 정보 나오게 구현

devfeijoa committed:
refactor: 너무 긴 이름을 가진 극장인 경우 말 줄임표로 나오게 구현

devfeijoa committed:
docs: README 업데이트

KwonDae commented on app/src/main/java/woowacourse/movie/view/reservation/seat/SeatSelectionContract.kt:
```diff
@@ -0,0 +1,88 @@
+    }
+}
+
+class SeatSelectionPresenter(
```
Presenter 구현체는 Contract 파일과 분리해주시는게 좋습니다.
수업자료에서 배웠듯이 Contract는 명세서 역할을 하기 때문이에요!

KwonDae commented on app/src/main/java/woowacourse/movie/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,152 @@
+            .setCancelable(false)
+            .setPositiveButton(R.string.reservation_dialog_positive) { _, _ ->
+                submitReservation()
+            }.setNegativeButton(R.string.reservation_dialog_negative) { dialog, _ -> dialog.dismiss() }
```
title, content는 언제든지 변할 수 있을거에요.
보다 유연하게 개선할 수 있을까요?

devfeijoa replied:
해당 피드백 반영하여 개선하였습니다 !
4375853a72e8ba14fc4591bfb00af5f856401f26

KwonDae commented on app/src/main/java/woowacourse/movie/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,152 @@
+        seat: Seat,
+        isSelected: Boolean,
+    ) {
+        findViewById<TableLayout>(R.id.tl_seat)
+            .children
+            .filterIsInstance<TableRow>()
+            .elementAtOrNull(seat.row)
+            ?.children
+            ?.filterIsInstance<TextView>()
+            ?.elementAtOrNull(seat.column)
```
updateSeatSelection이 자주 호출되는 함수라면
이 동작은 부담이 될 여지가 있어보여요.
어떻게 개선할 수 있을까요?

KwonDae commented on app/src/main/java/woowacourse/movie/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,152 @@
+            }
+    }
+
+    override fun showTotalPrice(price: Int) {
+        val tvPrice = findViewById<TextView>(R.id.tv_seat_select_total_price)
+        tvPrice.text = getString(R.string.reservation_total_money, price)
+    }
+
+    override fun showMovieTitle(title: String) {
+        findViewById<TextView>(R.id.tv_seat_select_movie_title).text = title
+    }
```
영화 정보와 관련된 데이터들을 개별적으로 세팅을 해줘야만 할까요?
이를테면
```kotlin
fun showMovieInfo(movie: Movie) {
 .. 
}
```

식으로 하나의 공통된 로직을 수행할 수 있지 않을까요?

KwonDae commented on app/src/main/java/woowacourse/movie/view/base/BaseActivity.kt:
```diff
@@ -0,0 +1,36 @@
+abstract class BaseActivity<T: ViewDataBinding>(
+    @LayoutRes private val layoutResId: Int,
+) : AppCompatActivity(layoutResId) {
+    private lateinit var binding: T
```
Base를 통해 Activity들의 binding 세팅 코드들을 줄여주는 역할이라고 생각되는데요
private 접근제한자를 사용하면 BaseActivity에서만 사용할 수 있는 구조가 아닐까요?

devfeijoa replied:
맞습니다...! 😓  해당 피드백 반영하여 수정하였습니다 ! 
c06ffb40bf7650103ada374838f82d0a3106168a

KwonDae commented on app/src/main/java/woowacourse/movie/view/extension/ParcelableExtension.kt:
```diff
@@ -0,0 +1,21 @@
+import android.os.Build
+import android.os.Bundle
+
+inline fun <reified T> Intent.getParcelableCompat(key: String): T =
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        this.getParcelableExtra(key, T::class.java) ?: throw IllegalArgumentException()
+    } else {
+        @Suppress("DEPRECATION")
+        getParcelableExtra(key) as? T ?: throw IllegalArgumentException()
+    }
```
만약 동료 개발자가 실수로 타입에 맞지 않은 값을 넣고 이를 인지하지 못한다면
사용자는 갑작스런 앱 종료를 마주하게 될거에요.

레벨1에서 배웠던 nullable을 활용할 수 있을까요?

KwonDae commented on app/src/main/java/woowacourse/movie/view/movies/MoviesPresenter.kt:
```diff
@@ -0,0 +1,24 @@
+            if ((index + 1) % 3 == 0) {
+                result.add(MovieListItem.AdItem(DummyAdvertisement.advertisement))
+            }
+        }
```
이 로직은 도메인 로직이 아닐까요?

KwonDae commented on app/src/main/java/woowacourse/movie/view/movies/cinema/CinemaSeclectionFragment.kt:
```diff
@@ -0,0 +1,67 @@
+                    }
+                },
+            )
+        list.layoutManager = LinearLayoutManager(context)
```
xml 에서도 이 옵션을 설정할 수 있습니다!

devfeijoa replied:
해당 피드백 반영하여 수정하였습니다 ! 
db27fd07611c268e8ee4b71fbf3943424b0aaa1e

KwonDae commented on app/src/main/java/woowacourse/movie/view/movies/cinema/CinemaSeclectionFragment.kt:
```diff
@@ -0,0 +1,67 @@
+            args.putParcelableArray(SCREENING_KEY, screenings.toTypedArray())
+            fragment.arguments = args
+            return fragment
+        }
```
```kotlin
fun newInstance(screenings: List<Screening>): CinemaSeclectionFragment {
   return CinemaSeclectionFragment().apply {
       arguments = bundleOf(..) 
   }    
}
```

식으로도 개선이 가능하겠네요 !

KwonDae commented on app/src/main/res/layout/activity_reservation.xml:
화면전환을 하면 사용자는 더이상 예매를 진행할 수 없습니다.
어떻게 개선할 수 있을까요?

devfeijoa replied:
해당 피드백 반영하여 수정하였습니다 !
aa5340f9f3e1fe73b7267628e84f140e660f903f

KwonDae commented on app/src/main/res/layout/activity_seat_selection.xml:
화면전환 대응이 필요해 보여요!

KwonDae commented on app/src/main/java/woowacourse/movie/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,152 @@
+    private fun setupViews(title: String) {
+        showMovieTitle(title)
+
+        findViewById<Button>(R.id.btn_seat_select_confirm).setOnClickListener {
+            presenter.showConfirmButton()
+        }
```
이러한 부분들이 말씀주신 step2의 데이터 바인딩 이벤트 처리를 적용할 부분입니다.
click 이벤트들에 전체적으로 적용해주세요!

KwonDae commented on app/src/main/java/woowacourse/movie/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,152 @@
+import woowacourse.movie.view.extension.getParcelableCompat
+import woowacourse.movie.view.reservation.result.ReservationResultActivity
+
+class SeatSelectionActivity :
```
화면 전환 이벤트가 발생하면
UI는 초기화 되지만,
내부 데이터는 남아 있어 정상 진행이 불가합니다.
이유는 무엇이고 어떻게 개선할 수 있는지 다음 요청때 공유해주세요!

devfeijoa replied:
이전 bundle에서 보낸 값과 presenter에서 사용하고 있는 값이 같은 값이여서 presenter에서 값을 업데이트 하면 두 개가 같이 업데이트 되는 문제가 발생하고 있었습니다 !

그래서 불변객체로 수정하여 개선하였습니다 ! 
21dc29b7ab4c3e9941f51390c351525ca759de34

KwonDae reviewed (changes requested):
이든 안녕하세요!
이번 영화 극장 선택 미션 함께할 제이든입니다.
주로 MVP 관련해서 코멘트를 남겨드렸고, 미처 구현하지 못한 부분은 다음 요청에 함께 반영 부탁드립니다.
천천히 확인해주시고, 어려운 부분은 편하게 말씀주세요!

devfeijoa committed:
refactor: Ktlint 적용

devfeijoa committed:
refactor: Presenter 분리

devfeijoa committed:
style: Ktlint 적용

devfeijoa committed:
feat: data binding 적용

devfeijoa committed:
refactor: project rename

devfeijoa commented:
@coderabbit pause

devfeijoa committed:
refactor: recyclerView 관련 네이밍 변경

lv -> rv

devfeijoa committed:
refactor: BaseActivity에서 private 접근제한자 제거

devfeijoa committed:
refactor: 네이밍 수정

devfeijoa committed:
docs: 2단계 기능 명세서 작성

devfeijoa committed:
refactor: 네이밍 변경

devfeijoa committed:
feat: BottomNavigationFragment.xml 구현

devfeijoa committed:
refactor: 네이밍 변경

devfeijoa committed:
refactor: 클래스 접근 제어자 변경

devfeijoa committed:
refactor: MainActivity 구현

devfeijoa committed:
refactor: main.xml 수정

devfeijoa committed:
feat: MovieListFragment 구현

devfeijoa committed:
refactor: 매직넘버 상수화

devfeijoa committed:
feat: SettingsFragment 구현

devfeijoa committed:
feat: ReservationHistoryFragment 구현

devfeijoa committed:
refactor: 네이밍 변경

devfeijoa committed:
refactor: MovieListFragment 호출

devfeijoa committed:
refactor: fragmentManager 변경

devfeijoa committed:
remove: 사용하지 않는 파일 제거

devfeijoa committed:
refactor: data binding 적용

devfeijoa committed:
refactor: data class로 변경

devfeijoa committed:
refactor: data binding 적용

devfeijoa committed:
refactor: dialog title, content를 받는 함수로 수정

devfeijoa committed:
refactor: data binding 적용

devfeijoa committed:
refactor: 불변 객체로 수정

이전 bundle에서 보낸 값과 presenter에서 사용하고 있는 값이 같은 값이여서 presenter에서 값을 업데이트 하면 두 개가 같이 업데이트 되는 문제 수정

devfeijoa committed:
refactor: xml에서 LinearLayoutManager 사용

KwonDae reviewed (approved):
이든 안녕하세요~!
다음 미션 진행을 위해 머지하도록 하겠습니다.
미처 반영하지 못한 피드백은 3,4 단계에서 함께 반영해주세요!
진행하면서 어려운 부분은 편하게 말씀주셔도 좋습니다 🙂
